File: Data Structures & Algorithms/count-number-of-islands/submission-1.java
class Solution {
    public int numIslands(char[][] grid) {
        Set<String> visited = new HashSet<>();
        int islands = 0;
        for(int row =0;row<grid.length; row++){
            for(int col=0; col < grid[0].length; col++){
                if(visited.contains(String.valueOf(row)+"-"+String.valueOf(col)) || grid[row][col] == '0'){
                    continue;
                }
                else{
                    dfs(grid,visited, row,col);
                    islands++;
                }
            }
        }
        return islands;
    }
    public void dfs(char[][] grid, Set<String> visited, int row, int col) {
        if(row < 0 || col < 0 || row >= grid.length || col >= grid[0].length){
            return;
        }
        String visitedStr = String.valueOf(row) +"-"+ String.valueOf(col);
        if(visited.contains(visitedStr) || grid[row][col] == '0'){
            return;
        }
        visited.add(visitedStr);
        dfs(grid,visited, row+1,col);
        dfs(grid,visited, row,col+1);
        dfs(grid,visited, row,col-1);
        dfs(grid,visited, row-1,col);
    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-1.cpp
class Solution {
    int max=0;
    int current=0;
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        for(int i =0 ;i < grid.size();i++){
            for(int j = 0; j < grid[0].size(); j++){
                if(grid[i][j] == 1){
                    current = 0;
                    dfs(grid, i, j);
                    if(current > max){
                        max = current;
                    }
                }
            }
        }
        return max;
    }

    void dfs(vector<vector<int>>& grid, int row, int col){
        
        if(row < 0 || col < 0 || row >= grid.size() || col >= grid[0].size()){
            return;
        }
        if(grid[row][col] == 0){
            return;
        }

        grid[row][col] = 0;
        current++;
        dfs(grid, row+1,col);
        dfs(grid, row,col-1);
        dfs(grid, row-1,col);
        dfs(grid, row,col+1);
    }
};
